src/Microsoft.Diagnostics.Monitoring.RestServer/MetricsService.cs

```diff
@@ -20,53 +22,86 @@ internal sealed class MetricsService : BackgroundService
+                    using IDisposable monitorListener = _optionsMonitor.OnChange((_, _) => optionsTokenSource.Cancel());
 
-                        await _counterPipeline.RunAsync(stoppingToken);
-                    }
-                    catch (Exception e) when (!(e is OperationCanceledException))
+                    EventPipeCounterGroup[] counterGroups = Array.Empty<EventPipeCounterGroup>();
+                    if (options.Providers.Count > 0)
                     {
-                        //Most likely we failed to resolve the pid. Attempt to do this again.
-                        if (_counterPipeline != null)
+                        //In the dotnet-monitor case, custom metrics are additive to the default counters.
+                        var eventPipeCounterGroups = new List<EventPipeCounterGroup>
                         {
-                            await _counterPipeline.DisposeAsync();
+                            new EventPipeCounterGroup { ProviderName = MonitoringSourceConfiguration.SystemRuntimeEventSourceName },
+                            new EventPipeCounterGroup { ProviderName = MonitoringSourceConfiguration.MicrosoftAspNetCoreHostingEventSourceName },
+                            new EventPipeCounterGroup { ProviderName = MonitoringSourceConfiguration.GrpcAspNetCoreServer }
+                        };
+
+                        foreach (MetricProvider customProvider in options.Providers)
+                        {
+                            var customCounterGroup = new EventPipeCounterGroup { ProviderName = customProvider.Provider };
+                            if (customProvider.Counters.Count > 0)
+                            {
+                                customCounterGroup.CounterNames = customProvider.Counters.ToArray();
+                            }
+                            eventPipeCounterGroups.Add(customCounterGroup);
                         }
-                        await Task.Delay(5000);
+                        counterGroups = eventPipeCounterGroups.ToArray();
+                    }
+
+                    _counterPipeline = new EventCounterPipeline(client, new EventPipeCounterPipelineSettings
+                    {
+                        CounterGroups = counterGroups,
+                        Duration = Timeout.InfiniteTimeSpan,
+                        RefreshInterval = TimeSpan.FromSeconds(options.UpdateIntervalSeconds)
+                    }, metricsLogger: new[] { new MetricsLogger(_store.MetricsStore) });
+
+                    using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken, optionsTokenSource.Token);
+                    await _counterPipeline.RunAsync(linkedTokenSource.Token);
+                }
+                catch (Exception e)
+                {
+                    if ((e is OperationCanceledException) &&
```

Comment:
Make this the when clause of the catch; could be something like:
`catch (Exception e) when (!IsCancellation(e, stoppingToken))`
...
`static bool IsCancellation(Exception e, CancellationToken token) { ... }`
I think this would be a little more readable and doesn't require rethrow.